### PR TITLE
Hidden layers arg fix + enabling linear encoders

### PR DESCRIPTION
**Pull request by cnellington**
`layers` now specifies the number of hidden layers in the encoders. (seems to have previously specified the number of dense connections, which was 2 by default with 1 hidden layer in between)

Users can now specify a linear encoder by using an MLP or NGAM with <1 layers, which defaults to a single matrix of trainable weights with no activations.

### CODE REVIEW

**Comment by cnellington** [comment]
link issue #97

File: contextualized/modules.py [hunk 1]
```diff
@@ -45,8 +45,11 @@ class MLP(nn.Module):
     """
     def __init__(self, input_dim, output_dim, width, layers, activation=nn.ReLU, link_fn=identity_link):
         super(MLP, self).__init__()
-        hidden_layers = lambda: [layer for _ in range(0, layers - 2) for layer in (nn.Linear(width, width), activation())]
-        mlp_layers = [nn.Linear(input_dim, width), activation()] + hidden_layers() + [nn.Linear(width, output_dim)]
+        if layers > 0:
+            hidden_layers = lambda: [layer for _ in range(0, layers - 1) for layer in (nn.Linear(width, width), activation())]
```
Comment (blengerich) [inline contextualized/modules.py:49]: Is `hidden_layers` used elsewhere? Seems a bit confusing to define a lambda function and only use it once.

**Review by blengerich — CHANGES_REQUESTED** [review]
One change requested. Otherwise, looks good.

File: contextualized/modules.py (hunk 1 above)
Comment (cnellington) [reply contextualized/modules.py:49]: Early in development there was an issue with a single nn.Linear being shared between multiple MLP modules and this was the fix, but I agree its bad practice. I'll radd the change if it passes the convergence tests.

**Review by blengerich — APPROVED** [review]
LGTM

**Review by juwaynilucman — APPROVED** [review]
LGTM

File: contextualized/modules.py [hunk 2]
```diff
@@ -45,8 +45,12 @@ class MLP(nn.Module):
     """
     def __init__(self, input_dim, output_dim, width, layers, activation=nn.ReLU, link_fn=identity_link):
         super(MLP, self).__init__()
-        hidden_layers = lambda: [layer for _ in range(0, layers - 2) for layer in (nn.Linear(width, width), activation())]
-        mlp_layers = [nn.Linear(input_dim, width), activation()] + hidden_layers() + [nn.Linear(width, output_dim)]
+        if layers > 0:
+            mlp_layers = [nn.Linear(input_dim, width), activation()] + \
+                    [layer for _ in range(0, layers - 1) for layer in (nn.Linear(width, width), activation())] + \
```
Comment (wtt102) [inline contextualized/modules.py:50]: Should the order here be reversed to this?

`[layer for layer in (nn.Linear(width, width), activation()) for _ in range(0, layers - 1)]`

File: contextualized/modules.py [hunk 3]
```diff
@@ -45,8 +45,12 @@ class MLP(nn.Module):
     """
     def __init__(self, input_dim, output_dim, width, layers, activation=nn.ReLU, link_fn=identity_link):
         super(MLP, self).__init__()
-        hidden_layers = lambda: [layer for _ in range(0, layers - 2) for layer in (nn.Linear(width, width), activation())]
-        mlp_layers = [nn.Linear(input_dim, width), activation()] + hidden_layers() + [nn.Linear(width, output_dim)]
+        if layers > 0:
+            mlp_layers = [nn.Linear(input_dim, width), activation()] + \
```
Comment (wtt102) [inline contextualized/modules.py:49]: Ben found before that the weight initializations for torch are slightly different (sqrt(1/n_features) vs 0.01)(https://pytorch.org/docs/stable/generated/torch.nn.Linear.html). Would we want to use this or separately initialize them?

**Review by wtt102 — COMMENTED** [review]
Looks good, just a couple things I wanted to clarify.

File: contextualized/modules.py (hunk 2 above)
Comment (blengerich) [reply contextualized/modules.py:50]: Probably should be rewritten for readability without a nested iterator, either way.

File: contextualized/modules.py (hunk 3 above)
Comment (blengerich) [reply contextualized/modules.py:49]: I think the default is okay for now (as long as we aren't mixing TF vs PyTorch, as you note), but we will want to allow custom inits in the future.

Comment (cnellington) [reply contextualized/modules.py:49]: Great comment! 

I agree with the new default. The only reason we would need the TF-Torch compatibility would be for reproducing the results in the NOTMAD preprint, but if we can reproduce those results with the new Torch version I think that may be less work. We can guarantee reproducibility within a torch version by using `torch.manual_seed(n)` and `numpy.random.seed(n)` prior to initializing a model and training. Between torch versions, we can use `contextualized.utils` to `save` and `load` weights. I can't think of a good reason to use custom weights otherwise, so seed-setting with model save/load load seems preferable, as long as we can reproduce the results in the NOTMAD pre-print. 

I can easily do this for the TCGA experiments once we merge the Torch NOTMAD #87 , @blengerich how difficult or neccessary do you think it is to reproduce the finance, sports, and SNARE-seq results in Torch?

File: contextualized/modules.py (hunk 2 above)
Comment (cnellington) [reply contextualized/modules.py]: Fixed, marking resolved

File: contextualized/modules.py (hunk 3 above)
Comment (blengerich) [reply contextualized/modules.py]: @cnellington We will want to reproduce the results using the Torch version, but this should be trivial now that @wtt102 's version is giving the same results as the TF version. Seeds will be important as well, I agree.

I believe this PR is ready to be merged as-is and we should make separate issues requesting addition of seed setting (maybe in `utils`) and custom inits.
